### PR TITLE
JENKINS-61956 ItemGroupMixin#createProject() does not call Jenkins#checkGoodName()

### PR DESCRIPTION
**Pull request by calvinpark**
https://issues.jenkins-ci.org/browse/JENKINS-61956

### CODE REVIEW

**Comment by res0nance** [comment]
https://javadoc.jenkins-ci.org/jenkins/model/Jenkins.html#checkGoodName-java.lang.String-
For reviewers here is the javadoc for `checkGoodName`

**Review by rsandell — CHANGES_REQUESTED** [review]
LGTM bu needs a test

**Comment by calvinpark** [comment]
@rsandell Can you please point me to which files I should look at for adding the tests? It's my first time dealing with Jenkins source so it'd be a lot faster with some hints. Thank you

**Review by res0nance — APPROVED** [review]

**Comment by daniel-beck** [comment]
Should be safe with multibranch projects given `NameEncoder` in `branch-api`, but should be explicitly confirmed.

**Comment by calvinpark** [comment]
@daniel-beck What's the process for checking the compatibility? I've tried `docker run -ti --rm -v maven-repo:/root/.m2 -v $(pwd)/out:/pct/out -v $(pwd)/war/target/jenkins.war:/pct/jenkins.war:ro -e ARTIFACT_ID=workflow-multibranch jenkins/pct` which failed

```
$ docker run -ti --rm -v maven-repo:/root/.m2 -v $(pwd)/out:/pct/out -v $(pwd)/war/target/jenkins.war:/pct/jenkins.war:ro -e ARTIFACT_ID=workflow-multibranch jenkins/pct
[...]
#############################################
##
## Starting to test plugin workflow-multibranch v2.21
## against MavenCoordinates[groupId=org.jenkins-ci.plugins, artifactId=plugin, version=2.234-SNAPSHOT]
##
#############################################
[...]
[INFO] --------< org.jenkins-ci.plugins.workflow:workflow-multibranch >--------
[INFO] Building Pipeline: Multibranch 2.22-SNAPSHOT
[...]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test (default-cli) on project workflow-multibranch: There are test failures.
[...]
+ cat /pct/tmp/work/workflow-multibranch/target/surefire-reports/2020-04-24T20-31-10_120-jvmRun1.dump
# Created at 2020-04-24T20:31:11.127
java.lang.NoClassDefFoundError: org/kohsuke/stapler/StaplerOverridable
[...]
```

**Comment by daniel-beck** [comment]
Not much of a PCT user, and unsure whether that configuration is even being tested there. I would just try the specific scenario manually.

**Comment by calvinpark** [comment]
Multibranch pipelines seem unaffected
image

```
println(Jenkins.instance.getItemByFullName("gl/new/branch"))
println(Jenkins.instance.getItemByFullName("gl/new%2Fbranch"))
```
returns
```
null
org.jenkinsci.plugins.workflow.job.WorkflowJob@ac925b6[gl/new%2Fbranch]
```

**Comment by calvinpark** [comment]
Any idea why `Deploy` stage is failing on CI?

**Comment by timja** [comment]
> Any idea why `Deploy` stage is failing on CI?

we're moving our agents to AWS and we have an intermittent connectivity issue, it won't block merging

**Comment by calvinpark** [comment]
Ready for review & merge

**Comment by calvinpark** [comment]
@rsandell Please review

**Comment by calvinatnvidia** [comment]
Please change the title of this PR to 
`JENKINS-61956 ItemGroupMixin#createProject() does not call Jenkins#checkGoodName`

**Comment by calvinatnvidia** [comment]
Actually please hold the merge. I'm finding that multibranch pipeline UI cashes if `/` is in the branch name. I'll experiment more to see how the previous test succeeded and report back.

**Comment by calvinatnvidia** [comment]
Running Jenkins with JDK 11 correctly displays the folder with `/` but running with JDK 8 crashes with error
```
javax.servlet.ServletException: org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/hudson/model/View/index.jelly:42:43: <st:include> org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/lib/hudson/projectView.jelly:84:48: <j:forEach> java.nio.CharBuffer.rewind()Ljava/nio/CharBuffer;
Caused by: org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/hudson/model/View/index.jelly:42:43: <st:include> org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/lib/hudson/projectView.jelly:84:48: <j:forEach> java.nio.CharBuffer.rewind()Ljava/nio/CharBuffer;
Caused by: java.lang.RuntimeException: org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/lib/hudson/projectView.jelly:84:48: <j:forEach> java.nio.CharBuffer.rewind()Ljava/nio/CharBuffer;
Caused by: org.apache.commons.jelly.JellyTagException: file:/Users/calvin/GitRepos/jenkins/core/src/main/resources/lib/hudson/projectView.jelly:84:48: <j:forEach> java.nio.CharBuffer.rewind()Ljava/nio/CharBuffer;
Caused by: java.lang.NoSuchMethodError: java.nio.CharBuffer.rewind()Ljava/nio/CharBuffer;
```

**Comment by calvinatnvidia** [comment]
The crashing behavior on JDK 8 is present on master branch without my changes. I don't believe this PR contributes to the crash.

**Comment by daniel-beck** [comment]
@calvinpark Jenkins supports building on exactly JDK 8 and JDK 11. My guess is you built it on a different major version.

Also, it looks like you removed intermediate elements from the track trace, please don't do that.

**Comment by MarkEWaite** [comment]
@calvinpark I think that no such method exception is the same exception that was reported in [JENKINS-61105](https://issues.jenkins-ci.org/browse/JENKINS-61105) when a Jenkins release candidate was inadvertently compiled with Java 11 and run with Java 8.  You may need to `mvn clean` to switch from Java 11 to Java 8 for testing.

**Comment by calvinpark** [comment]
@daniel-beck & @MarkEWaite, you were right. The error was due to compiling with JDK 11 then running it with JDK 8. It ran successfully and the multibranch pipelines with `/` in the name displayed correctly when it was compiled and run with JDK 8.

File: test/src/test/java/hudson/model/ItemGroupMixInTest.java
Comment (MarkEWaite) [inline test/src/test/java/hudson/model/ItemGroupMixInTest.java:267]: JUnit assertThrows is now available (JUnit 4.13) to make these types of assertion checks cleaner.  

I'm not clear on the catch(Failure), since the class Failure seems to be a JUnit class, not an exception class.

Comment (calvinpark) [reply test/src/test/java/hudson/model/ItemGroupMixInTest.java:267]: Would you like to me to change it to use `assertThrows`?

Comment (calvinpark) [reply test/src/test/java/hudson/model/ItemGroupMixInTest.java]: Changed

**Review by MarkEWaite — APPROVED** [review]
Looks great.  Thanks!

**Comment by MarkEWaite** [comment]
@rsandell are the added tests enough to resolve your concern?

**Comment by calvinpark** [comment]
A friendly ping :)

**Comment by MarkEWaite** [comment]
@rsandell can you confirm your concern was addressed by the added test?  The [merge process](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#common-merge-process)  asks that there be no outstanding requests for change before we start the 24 hour merge clock.  Yours is an outstanding request for change.

**Comment by calvinpark** [comment]
@rsandell doesn't seem available. Is there a way to have someone else respond on his behalf?

**Comment by MarkEWaite** [comment]
I've added the "ready for merge" label and will merge in 24 hours, assuming no objections from others.  If someone objects, they need to express those concerns here.

**Comment by calvinpark** [comment]
Whoa. How did the CI pass?

**Comment by jeffret-b** [comment]
It's weird. I don't know how it happened, but it's working now.

**Comment by calvinpark** [comment]
Glad to hear. Thank you.

**Comment by jglick** [comment]
> Multibranch pipelines seem unaffected

FTR see discussion in https://github.com/jenkinsci/branch-api-plugin/pull/200.

**Comment by Kolky** [comment]
Possibly the cause for: https://issues.jenkins.io/browse/JENKINS-64121

**Pull request merged**
